### PR TITLE
Yet another dedescriptorization step

### PR DESCRIPTION
**Pull request by homuroll**

### CODE REVIEW

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/descriptors/DescriptorUtils.kt [hunk 1]
```diff
@@ -77,6 +78,14 @@ internal fun IrSimpleFunction.resolveFakeOverride(allowAbstract: Boolean = false
     return realSupers.first { allowAbstract || it.modality != Modality.ABSTRACT }
 }
 
+internal fun IrSimpleFunction.overrides(superFunctionSymbol: IrSimpleFunctionSymbol): Boolean {
+    if (superFunctionSymbol in overriddenSymbols) return true
+    return overriddenSymbols.any { it.owner.overrides(superFunctionSymbol) }
```
Comment (SvyatoslavScherbina) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/descriptors/DescriptorUtils.kt]: The implementation is kinda exponential time in worst case.
I believe general utility method shouldn't be like this.
Its usage in `InteropLowering.kt` can be replaced by a simple name-based check, so we probably can avoid adding this method at all for now.

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/descriptors/DescriptorUtils.kt [hunk 2]
```diff
@@ -254,6 +263,11 @@ fun IrConstructorCall.getStringValue(name: String): String {
     return (getValueArgument(parameter.index) as IrConst<String>).value
 }
 
+fun <T> IrConstructorCall.getArgumentValueOrNull(name: String): T? {
```
Comment (SvyatoslavScherbina) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/descriptors/DescriptorUtils.kt]: I guess the name of this function should clearly indicate that it is applicable to annotations only.

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/InteropLowering.kt [hunk 3]
```diff
@@ -839,11 +834,11 @@ private class InteropTransformer(val context: Context, override val irFile: IrFi
 
         expression.transformChildrenVoid(this)
         builder.at(expression)
-        val descriptor = expression.descriptor.original
         val function = expression.symbol.owner
 
-        if (descriptor == interop.nativePointedRawPtrGetter ||
-                OverridingUtil.overrides(descriptor, interop.nativePointedRawPtrGetter, false)) {
+        // TODO: this is fragile.
```
Comment (SvyatoslavScherbina) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/InteropLowering.kt]: Yes it is.
What I meant is to check here that class derives from `NativePointed` and that this function is getter of property named `rawPtr`.
In this case this property overrides `NativePointed.rawPtr`.

Btw, it is possible that overriding check is not required anymore, after this patch: https://github.com/JetBrains/kotlin-native/commit/2b00850164d6aa5680248ac9ed7f3d35ec5ef1f9#diff-87031cea584971a666c72def5450f827R27

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/InteropLowering.kt [hunk 4]
```diff
@@ -836,9 +837,8 @@ private class InteropTransformer(val context: Context, override val irFile: IrFi
         builder.at(expression)
         val function = expression.symbol.owner
 
-        // TODO: this is fragile.
-        if (function.isGetter && function.parent is IrClass
-                && function.name == symbols.interopNativePointedRawPtrGetter.owner.name) {
+        if ((function as? IrSimpleFunction)?.resolveFakeOverrideMaybeAbstract()?.symbol
+                == symbols.interopNativePointedGetRawPointer) {
```
Comment (SvyatoslavScherbina) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/InteropLowering.kt]: `interopNativePointedRawPtrGetter`.

Comment (homuroll) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/InteropLowering.kt]: 🤦‍♂

**Pull request merged**